### PR TITLE
Introduce and pass vmmops Trait to hypervisor module

### PR DESCRIPTION
**Pull request by praveen-pk**
This pull request introduces Vmmops Trait, which is implemented by Vm Object in VMM module and passed to Vm and Vcpu objects in hypervisor module.

Fixes: #1660

### CODE REVIEW

File: hypervisor/src/kvm/mod.rs [hunk 1]
```diff
@@ -330,6 +341,16 @@ impl vm::Vm for KvmVm {
...
     }
+
+    // set vmmops interface
+    fn set_vmmops(&mut self, vmmops: Arc<Mutex<dyn VmmOps>>) -> vm::Result<()>{
+        self.vmmops = Some(vmmops);
+        Ok(())
+    }
+    // get vmmops interface
+    fn get_vmmops(&self) -> vm::Result<Arc<Mutex<dyn VmmOps>>>{
+        Ok(self.vmmops.clone().unwrap())
```
Comment (liuw) [inline hypervisor/src/kvm/mod.rs:352]: Please don't `unwrap` unconditionally. This function returns an `Result` so we should be able to pass the error up.

File: hypervisor/src/vm.rs [hunk 2]
```diff
@@ -117,6 +118,26 @@ pub enum HypervisorVmError {
...
+    ///
+    /// Read Guest memory
+    ///
+    #[error("Failed to read guest memory: {0}")]
+    GuestMemRead(#[source] anyhow::Error),
+    ///
+    /// Read from MMIO Bus
+    ///
+    #[error("Failed to read from MMIO Bus: {0}")]
+    MMioBusRead(#[source] anyhow::Error),
```
Comment (liuw) [inline hypervisor/src/vm.rs:135]: `MmioBusRead` here.

File: hypervisor/src/vm.rs [hunk 3]
```diff
@@ -117,6 +118,26 @@ pub enum HypervisorVmError {
...
+    ///
+    /// Write to MMIO Bus
+    ///
+    #[error("Failed to write to MMIO Bus: {0}")]
+    MMioBusWrite(#[source] anyhow::Error),
```
Comment (liuw) [inline hypervisor/src/vm.rs:140]: `MmioBusWrite` here.

File: hypervisor/src/kvm/mod.rs [hunk 4]
```diff
@@ -403,19 +424,21 @@ impl hypervisor::Hypervisor for KvmHypervisor {
...
                 msrs,
                 state: VmState {},
-            }))
+            })))
         }
 
         #[cfg(any(target_arch = "arm", target_arch = "aarch64"))]
         {
-            Ok(Arc::new(KvmVm {
+            Ok(Arc::new(Mutex::new((KvmVm {
```
Comment (liuw) [inline hypervisor/src/kvm/mod.rs:437]: There is no need to add an extra pair of parentheses.

File: pci/src/vfio.rs [hunk 5]
```diff
@@ -555,15 +555,15 @@ impl VfioPciDevice {
                 }
 
                 let slot = mem_slot();
-                let mem_region = vm.make_user_memory_region(
+                let mem_region = vm.lock().unwrap().make_user_memory_region(
```
Comment (liuw) [inline pci/src/vfio.rs:558]: Add `let vm = vm.lock().unwrap();` at the beginning of the function then you don't need to modify the rest of the function.

Actually I think this makes more sense because you will want the lock to be held for the entirety of this function, no?

File: vmm/src/lib.rs [hunk 6]
```diff
@@ -329,35 +329,36 @@ impl Vmm {
         }
 
         // Now we can boot the VM.
-        if let Some(ref mut vm) = self.vm {
-            vm.boot()
+        if self.vm.is_some()
```
Comment (liuw) [inline vmm/src/lib.rs:332]: There is no need to change to this new pattern, is there?

`self.vm` is still going to be of either `Some(X)` or `None`.

You can shorten the following line quite a bit. That makes things easier to reason about.

File: vmm/src/vm.rs [hunk 7]
```diff
@@ -489,19 +490,20 @@ impl Vm {
...
         // of the regular code path creating everything from scratch.
         new_vm
+            .lock().unwrap()
             .device_manager
             .lock()
             .unwrap()
             .create_devices()
             .map_err(Error::DeviceManager)?;
-
+
```
Comment (liuw) [inline vmm/src/vm.rs:506]: Seems to be unnecessary white space change here.

File: hypervisor/src/kvm/mod.rs [hunk 8]
```diff
@@ -160,8 +164,15 @@ impl vm::Vm for KvmVm {
             .fd
             .create_vcpu(id)
             .map_err(|e| vm::HypervisorVmError::CreateVcpu(e.into()))?;
+        let vmmops = {
+            if self.vmmops.is_some(){
+                Some(self.vmmops.as_ref().unwrap().clone())
+            }
+            else{   None    }
+        };
```
Comment (sameo) [inline hypervisor/src/kvm/mod.rs:172]: This would be more idiomatic:

```Rust
let vmmops = self.vmmops.as_ref().map(|ops| ops.clone());
```

File: hypervisor/src/kvm/mod.rs [hunk 9]
```diff
@@ -160,8 +164,15 @@ impl vm::Vm for KvmVm {
...
         let vcpu = KvmVcpu {
             fd: vc,
+            vmmops: vmmops,
```
Comment (sameo) [inline hypervisor/src/kvm/mod.rs:175]: Just `vmmops,`

**Review by sameo — COMMENTED** [review]
Thanks a lot for the PR @praveen-pk 

Please see the comments. Also, you should squash all 3 first commits into a single one that introduces the VmmOps trait, with a description on what we're trying to solve with this trait and how we're solving it.

Also, please run the `dev_cli.sh test --cargo` command, it will probably come back with a few errors that you'll need to fix in order to pass a CI run.

File: pci/src/vfio.rs (hunk 5 above)
Comment (praveen-pk) [reply pci/src/vfio.rs]: Added a local variable to the top of the function, made sense to acquire a single lock to vm at the top of the function and reuse it.

File: vmm/src/lib.rs [hunk 10]
```diff
@@ -556,8 +557,9 @@ impl Vmm {
     }
 
     fn vm_add_disk(&mut self, disk_cfg: DiskConfig) -> result::Result<Vec<u8>, VmError> {
-        if let Some(ref mut vm) = self.vm {
-            let info = vm.add_disk(disk_cfg).map_err(|e| {
+        if self.vm.is_some() {
+            let info = self.vm.as_ref().unwrap().clone().lock().unwrap()
+            .add_disk(disk_cfg).map_err(|e| {
```
[the remark this answers is not in the document]
Comment (praveen-pk) [reply vmm/src/lib.rs]: I updated this access pattern because, the new version felt more readable, reverted

**Comment by praveen-pk** [comment]
Still fixing issues reported by `dev_cli.sh tests --cargo`.

**Comment by liuw** [comment]
> 
> 
> An issue reported by CI run is following:
> 
> ```
> error: field is never read: `vmmops`
>    --> hypervisor/src/kvm/mod.rs:491:5
>     |
> 491 |     vmmops: Option<Arc<Mutex<dyn vm::VmmOps>>>,
>     |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>     |
>     = note: `-D dead-code` implied by `-D warnings`
> ```
> 
> vmmops in KVmVm is never really used. Any suggestions on how I should handle this failure? Does adding a feature guard around vmmops (disabled for KVM) make sense?

You should switch KVM to use VmmOps. And then perhaps we can drop some of the VmExit variants. CH doesn't need to have two paths for handling requests from the hypervisor.

This is just my idea anyway. Please wait for Samuel & co to chime in.

**Comment by sameo** [comment]
> An issue reported by CI run is following:
> 
> ```
> error: field is never read: `vmmops`
>    --> hypervisor/src/kvm/mod.rs:491:5
>     |
> 491 |     vmmops: Option<Arc<Mutex<dyn vm::VmmOps>>>,
>     |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>     |
>     = note: `-D dead-code` implied by `-D warnings`
> ```
> 
> vmmops in KVmVm is never really used. Any suggestions on how I should handle this failure? Does adding a feature guard around vmmops (disabled for KVM) make sense?

We really should have the kvm implementation making use of those VmmOps and offload all the IO/MMIO vm exit handling to the kvm implementation in the hypervisor crate.

If there's no VmmOps set, then we should return an error from kvm.

That brings the next open, that we should leave for some further PR: We are moving to an architecture where we mandate any `hypervisor::Vm` implementation to also implement `VmmOps` and although I wish there was an idiomatic way to express that with trait bounds, I have not yet found a proper way to do so.

File: pci/src/vfio.rs [hunk 11]
```diff
@@ -584,15 +589,15 @@ impl VfioPciDevice {
                 let (mmap_offset, _) = self.device.get_region_mmap(region.index);
 
                 // Remove region
-                let r = self.vm.make_user_memory_region(
+                let r = self.vm.lock().unwrap().make_user_memory_region(
```
Comment (liuw) [inline pci/src/vfio.rs:592]: Use `let vm = vm.lock().unwrap()` at the beginning. Please check all other places where the same pattern can be used.

Comment (praveen-pk) [reply pci/src/vfio.rs]: Checked and fixed all such occurrences.  I only fixed instances where moving the initialization to top, would improve performance.  I left all single invocation as is.

**Comment by praveen-pk** [comment]
```./scripts/dev_cli.sh tests --unit``` passes with the latest revision.

```./scripts/dev_cli.sh tests --cargo``` fails with the following error, working on fixing it:

```
+ cargo fmt -- --check
error: attributes are not yet allowed on `if` expressions
   --> /cloud-hypervisor/vmm/src/memory_manager.rs:995:9
    |
995 |         #[cfg(target_arch = "x86_64")]
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^


real    0m0.779s
user    0m0.759s
sys     0m0.019s
```

File: hypervisor/src/kvm/mod.rs [hunk 12]
```diff
@@ -696,9 +696,35 @@ impl cpu::Vcpu for KvmVcpu {
...
                 #[cfg(target_arch = "x86_64")]
-                VcpuExit::IoIn(addr, data) => Ok(cpu::VmExit::IoIn(addr, data)),
+                VcpuExit::IoIn(addr, data) => {
+                    if let Some(ref vmmops) = self.vmmops {
+                        let ret = vmmops.lock().unwrap().pio_read(addr.into(), data);
+                        if ret.is_ok() {
+                            return Ok(cpu::VmExit::Ignore);
+                        }
+                    }
+                    Err(cpu::HypervisorCpuError::RunVcpu(anyhow!(
```
Comment (praveen-pk) [inline hypervisor/src/kvm/mod.rs:706]: In the original handler, a warning is thrown if an IO address cannot be read/written to. While using vmmops I modified it to return an error. With this change unit tests seem to pass just fine. 

Can someone confirm if (or not) warning on read/write is the correct approach here?

File: hypervisor/src/kvm/mod.rs [hunk 13]
```diff
@@ -347,7 +347,7 @@ impl vm::Vm for KvmVm {
     ///
     /// Set the VM state
     ///
-    fn set_state(&self, _state: &VmState) -> vm::Result<()> {
+    fn set_state(&self, _state: VmState) -> vm::Result<()> {
```
Comment (liuw) [inline hypervisor/src/kvm/mod.rs:350]: OOI What did clippy warn about using a reference?

File: option_parser/src/lib.rs [hunk 14]
```diff
@@ -189,7 +189,7 @@ impl FromStr for IntegerList {
             let items: Vec<&str> = range.split('-').collect();
 
             if items.len() > 2 {
-                return Err(IntegerListParseError::InvalidValue(range.to_string()));
+                return Err(IntegerListParseError::InvalidValue((*range).to_string()));
```
Comment (liuw) [inline option_parser/src/lib.rs:192]: Why is this file changed? This has nothing to do with VmmOps, right?

Comment (liuw) [reply option_parser/src/lib.rs:192]: Oh, so you attempted to fix clippy warnings in this PR as well.

I think those warnings should be fixed in a separate PR.

File: hypervisor/src/kvm/mod.rs [hunk 15]
```diff
@@ -696,9 +696,35 @@ impl cpu::Vcpu for KvmVcpu {
...
+                        addr
+                    )))
+                }
                 #[cfg(target_arch = "x86_64")]
-                VcpuExit::IoOut(addr, data) => Ok(cpu::VmExit::IoOut(addr, data)),
+                VcpuExit::IoOut(addr, data) => {
+                    if let Some(ref vmmops) = self.vmmops {
+                        let ret = vmmops.lock().unwrap().pio_write(addr.into(), data);
+                        if ret.is_ok() {
+                            if addr == DEBUG_IOPORT && data.len() == 1 {
```
Comment (liuw) [inline hypervisor/src/kvm/mod.rs:716]: The special handling of DEBUG_IOPORT should not be here. It should be in `pio_write`.

Reasons:
1. It is not specific to KVM.
2. It helps you reduce the amount of code churn -- you won't need to define DEBUG_IOPORT in `hypervisor` and at the same time introduce a `LogDebugPort`.

File: hypervisor/src/kvm/mod.rs (hunk 12 above)
Comment (liuw) [reply hypervisor/src/kvm/mod.rs:706]: It depends on whether you want the VCPU to get terminated or not in that situation.

Since your goal doesn't not include changing the failure mode, I would suggest you retain the original behaviour. That is, only warn but don't return an error.

**Comment by liuw** [comment]
I have some further comments regarding the commit messages, mostly to keep things consistent.

1. Please add a space between component names. Please write `hypervisor, vmm` instead of `hypervisor,vmm`.
2. Please use `VmmOps` through out the subject line and commit message.
3. There is no need to capitalize `Trait` in the first commit's subject line.

File: hypervisor/src/kvm/mod.rs (hunk 12 above)
Comment (sameo) [inline hypervisor/src/kvm/mod.rs:706]: I think we want the following logic instead:

```Rust
                VcpuExit::IoIn(addr, data) => {
                    if let Some(ref vmmops) = self.vmmops {
                        return vmmops
                            .lock()
                            .unwrap()
                            .pio_read(addr.into(), data)
                            .map(|_| cpu::VmExit::Ignore)
                            .map_err(|e| cpu::HypervisorCpuError::RunVcpu(e.into()));
                    }

                    Ok(cpu::VmExit::IoIn(addr, data))
                }
```

1. If we have a `VmmOps`, we handle the VM exit through it and return the ops result itself.
1. If we don't have a `VmmOps`, we offload the VM exit back to the VMM.

We could argue about the second step and make it return an error when it's missing a `VmmOps`, but since we don't have a way to make it mandatory for now, we should gracefully handle this option.

On Cloud Hypervisor, we would error when receiving a `VmExit::IoIn` from the hypervisor crate.

File: option_parser/src/lib.rs (hunk 14 above)
Comment (liuw) [reply option_parser/src/lib.rs:192]: Okay, these show up when running `dev_cli.sh tests --all`.

The fixes here should really be in a separate PR. They don't need to wait until `VmmOps` is settled.

File: hypervisor/src/kvm/mod.rs (hunk 15 above)
Comment (praveen-pk) [reply hypervisor/src/kvm/mod.rs:716]: log_debug_ioport was not available via vmmops and so I thought I could pass the control back to VmExit for invoking log_debug_ioport.

Thanks for clarifying DEBUG_IOPORT is specific to CH and not KVM. I will move it back into pio_write and figure out piping for log_debug_ioport.

**Comment by praveen-pk** [comment]
Moved all clippy fixes to https://github.com/cloud-hypervisor/cloud-hypervisor/pull/1761. Will address above feedback with my next updates.

File: hypervisor/src/kvm/mod.rs (hunk 12 above)
Comment (praveen-pk) [reply hypervisor/src/kvm/mod.rs:706]: make sense Samuel, will adopt this logic

File: vmm/src/cpu.rs [hunk 16]
```diff
@@ -418,7 +372,10 @@ impl Vcpu {
                     }
                     Ok(true)
                 }
-
+                VmExit::LogDebugPort(data) => {
```
Comment (liuw) [inline vmm/src/cpu.rs:375]: This won't build.

File: hypervisor/src/vm.rs [hunk 17]
```diff
@@ -184,4 +215,15 @@ pub trait Vm: Send + Sync {
...
+    fn set_vmmops(&mut self, vmmops: Arc<Mutex<dyn VmmOps>>) -> vm::Result<()>;
+}
+
+pub trait VmmOps: Send + Sync {
+    fn guest_mem_write(&self, buf: &[u8], gpa: u64) -> vm::Result<usize>;
+    fn guest_mem_read(&self, buf: &mut [u8], gpa: u64) -> vm::Result<usize>;
+    fn mmio_read(&self, addr: u64, data: &mut [u8]) -> vm::Result<()>;
+    fn mmio_write(&self, addr: u64, data: &[u8]) -> vm::Result<()>;
+    fn pio_read(&self, addr: u64, data: &mut [u8]) -> vm::Result<()>;
+    fn pio_write(&self, addr: u64, data: &[u8]) -> vm::Result<()>;
```
Comment (sboeuf) [inline hypervisor/src/vm.rs:228]: @praveen-pk @liuw 
Sorry to comment a bit late on this PR, but here is how I think we could make this change less intrusive:
I would keep the `VmmOps` trait but I would remove both `guest_mem_write()` and `guest_mem_read()` from it. Instead, we could expect the `hypervisor` to rely on `vm-memory`, which means we could set/pass a `GuestMemoryAtomic<GuestMemoryMmap>`, therefore the `hypervisor` crate could directly deal with read/write from/to the guest memory.
This would simplify the `VmmOps` trait, but the big benefit of this approach would be to implement the `VmmOps` from the `DeviceManager` directly instead of `Vm`. I'm really not a huge fan of having the entire `Vm` object being locked every time an operation from the `VmmOps` is performed.

@rbradford @sameo WDYT?

File: vmm/src/vm.rs [hunk 18]
```diff
@@ -1540,6 +1544,58 @@ impl Pausable for Vm {
...
+    fn pio_read(&self, addr: u64, data: &mut [u8]) -> hypervisor::vm::Result<()> {
+        self.device_manager
+            .lock()
+            .unwrap()
+            .io_bus()
+            .read(addr, data)
+            .map_err(|e| HypervisorVmError::IoBusRead(e.into()))
+    }
+
+    fn pio_write(&self, addr: u64, data: &[u8]) -> hypervisor::vm::Result<()> {
```
Comment (sameo) [inline vmm/src/vm.rs:1589]: This here is where you should implement the debug port handling and logging:

```Rust
if addr == DEBUG_IOPORT && data.len() == 1 {
     self.log_debug_ioport(data);
}
```

File: hypervisor/src/vm.rs (hunk 17 above)
Comment (sameo) [reply hypervisor/src/vm.rs:228]: @sboeuf Since the hypervisor crate already depends on the `vm-memory` crate, that would make sense.
What should be passed to the hypervisor crate is something more generic than `GuestMemoryAtomic<GuestMemoryMmap>` though. I don't think we should expect the hypervisor crate could modify the guest address space and could only pass something that implements `GuestAddressSpace`?

Comment (sboeuf) [reply hypervisor/src/vm.rs:228]: I think if we don't pass a `GuestMemoryAtomic<>`, we might run into the issue that some memory regions might have been added, hence the hypervisor would not have access to the updated `GuestMemoryMmap`.
And the only method from the `GuestAddressSpace` trait is `memory()`, which is defined as follows:
```
    /// Return an object (e.g. a reference or guard) that can be used
    /// to access memory through this address space.  The object provides
    /// a consistent snapshot of the memory map.
    fn memory(&self) -> Self::T;
```
So the problem is that it simply provides a snapshot at a given time, but I'm worried we might break things for memory hotplug.

Comment (russell-islam) [reply hypervisor/src/vm.rs:228]: If we don't agree implementing vmmops in the Vm we have to rewrite the patches almost completely. It would be better if we have a consensus ASAP.

Comment (sameo) [reply hypervisor/src/vm.rs:228]: > I think if we don't pass a `GuestMemoryAtomic<>`, we might run into the issue that some memory regions might have been added, hence the hypervisor would not have access to the updated `GuestMemoryMmap`.
> And the only method from the `GuestAddressSpace` trait is `memory()`, which is defined as follows:
> 
> ```
>     /// Return an object (e.g. a reference or guard) that can be used
>     /// to access memory through this address space.  The object provides
>     /// a consistent snapshot of the memory map.
>     fn memory(&self) -> Self::T;
> ```
> 
> So the problem is that it simply provides a snapshot at a given time, but I'm worried we might break things for memory hotplug.

So we're forcing the VMM to use an hotpluggable memory implementation, which is not something that's portable. Then I think it makes sense to keep those as part of the VmmOps trait and let the VMM handle memory operations depending on how it decided to implement its guest address space.

Comment (sboeuf) [reply hypervisor/src/vm.rs:228]: Well I understand your frustration, but it's important that we merge the right design the first time, otherwise this will cost everybody more time if we decide to modify it later. Moreover, I actually think the proposal would simplify the PR.
@rbradford WDYT?

Comment (sboeuf) [reply hypervisor/src/vm.rs:228]: @sameo 
>So we're forcing the VMM to use an hotpluggable memory implementation, which is not something that's portable.

Ok that's fair, even if I think `GuestMemoryAtomic<>` should be the smallest unit, even for VMM without hotplug needs.

I would still like to avoid locking the entire `Vm` structure. So how about we define a new structure which implements `VmmOps` and which has a reference to both `MemoryManager` and `DeviceManager`? This structure does not need to be shared between the `vmm` and `hypervisor` crate, therefore it would greatly simplify the whole `Mutex` complexity.
Something like
```rust
struct VmOps {
    memory_manager: ...
    device_manager: ...
}
impl VmmOps for VmOps {
...
}
```

Comment (russell-islam) [reply hypervisor/src/vm.rs:228]: > > If we don't agree implementing vmmops in the Vm, we have to rewrite the patches almost completely. It would be better if we have an agreement ASAP.
> 
> Well I understand your frustration, but it's important that we merge the right design the first time, otherwise this will cost everybody more time if we decide to modify it later. Moreover, I actually think the proposal would simplify the PR.
> @rbradford WDYT?

I can't agree more. I am not frustrated thought. :)

Comment (sboeuf) [reply hypervisor/src/vm.rs:228]: @russell-islam we talked about this offline with @sameo and @rbradford. I'm gonna have a quick try at implementing the `VmOps` solution mentioned above. I'll get back to you with when ready.

Comment (russell-islam) [reply hypervisor/src/vm.rs:228]: > @russell-islam we talked about this offline with @sameo and @rbradford. I'm gonna have a quick try at implementing the `VmOps` solution mentioned above. I'll get back to you when ready.

Sounds good. :+1:

Comment (sboeuf) [reply hypervisor/src/vm.rs:228]: @russell-islam here is the simplified version I was thinking about: https://github.com/cloud-hypervisor/cloud-hypervisor/pull/1793
I'm just missing a few things regardind the VM exits (moving things from `vmm/src/cpu.rs` to `hypervisor/src/kvm/mod.rs`) but the point is to show how I got rid of the locks. Let me know what you think about it.

Comment (sboeuf) [reply hypervisor/src/vm.rs:228]: /cc @sameo @rbradford

**Comment by rbradford** [comment]
Superseded by #1793
